### ISSUE TITLE
lazy/check: broken logic

### ISSUE THREAD

**Issue by airwoodix** [qtile/qtile#1646]
# Issue description

I was too quick in testing for #1645. The patch fixes the reported issue (#1629) but [the logic](https://github.com/qtile/qtile/blob/8d5c17354cbfa58cd0b4e58b3cbe245e225fb66f/libqtile/lazy.py#L77) in ``LazyCall.check`` seems broken to me. Sorry for the noise.

The code reads:

```python
    def check(self, q) -> bool:
        if self._layout is not None:
            if self._layout == 'floating':
                if q.current_window.floating:
                    return True
                return False
            if q.current_layout.name != self._layout:
                if q.current_window and q.current_window.floating and not self._when_floating:
                    return False
        return True
```

which implies that the call is evaluated even if ``q.current_layout.name != self._layout`` when e.g. the current window is not floating.

The [old logic](https://github.com/qtile/qtile/blob/a3ae3c623859b247813fc1876b188c4d40e84df0/libqtile/command.py#L293) looks better but has an edge case when the current layout is ``floating`` but there's no window any more (access ``.floating`` on ``q.current_window = None``). This is also the case in the code above.

**Comment by ramnes** [comment]
Can you describe why it is broken, exactly? What is it that you try, what is the expected result, and what do you get?

Sorry if that's obvious but we have numerous issues to deal with and it's hard to give time for each one of them when it's not ELI5.

**Comment by airwoodix** [comment]
I updated the the issue description above.
